File: seleniumCSharp/SeleniumNunitPractice/HuynhHoangTram/Page/LoginPage.cs
using System;
using System.Threading;
using System.Linq;
using HuynhHoangTram.Core;
using OpenQA.Selenium;
namespace HuynhHoangTram.Page;

public class LoginPage
{
    private WebObject _userNameInput = new WebObject(By.CssSelector("input#username"), "Username Input");

    private WebObject _passwordInput = new WebObject(By.CssSelector("input#password"), "Password Input");

    private WebObject _loginBtn = new WebObject(By.XPath("//input[@value='Login']"), "Login button");

    private WebObject _errorMessage = new WebObject(By.CssSelector("form[name='loginForm'] div.alert-danger"),"Error Message");
    private WebObject _usernamerequiredmessage = new WebObject(By.XPath("//div[@ng-messages='submitted && loginForm.username.$error']"), "Username Required Message");

    private WebObject _passwordrequiredmessage = new WebObject(By.XPath("//div[@ng-messages='submitted && loginForm.password.$error']"), "Password Required Message");

    public LoginPage()
    {
    }

    public void EnterUsername(string username)
    {
        _userNameInput.EnterText(username);
    }

    public void EnterPassword(string password)
    {
        _passwordInput.EnterText(password);
    }

    public void ClickLogin()
    {
        _loginBtn.ClickOnElement();
    }

     public bool IsLoginFailed()
    {
        return WebObjectExtension.WaitUntilAnyElementDisplayed(
            _errorMessage,
            _usernamerequiredmessage,
            _passwordrequiredmessage
        );
    }
    public void Login(string username, string password)
    {
        EnterUsername(username);

        EnterPassword(password);

        ClickLogin();
    }
}

File: playwright/practice/HoangTramHuynh-PW/page-object/login-page.ts
import { expect, Locator, Page } from '@playwright/test';
import { BARE_URL } from '../constant/url';

export class LoginPage {
  private readonly page: Page;
  private readonly userNameInput: Locator;
  private readonly passwordInput: Locator;
  private readonly loginBtn: Locator;
  private readonly loginErrorMessage: Locator;

  constructor(page: Page) {
    this.page = page;

    this.userNameInput = page.locator('input#username');
    this.passwordInput = page.locator('input#password');
    this.loginBtn = page.getByRole('button', { name: 'Login' });

    this.loginErrorMessage = page
      .locator(
        "form[name='loginForm'] div.alert-danger, " +
        "form[name='loginForm'] [role='alert'], " +
        "form[name='loginForm'] div[ng-messages]"
      )
      .filter({ hasText: /\S/ });
  }

  async openLoginPage(): Promise<void> {
    await this.page.goto(BARE_URL);
  }

  async enterUsername(username: string): Promise<void> {
    await this.userNameInput.fill(username);
  }

  async enterPassword(password: string): Promise<void> {
    await this.passwordInput.fill(password);
  }

  async clickLogin(): Promise<void> {
    await this.loginBtn.click();
  }

  async login(username: string, password: string): Promise<void> {
    await this.enterUsername(username);
    await this.enterPassword(password);
    await this.clickLogin();
  }

  async verifyLoginFailedMessageIsDisplayed(): Promise<void> {
    await expect(this.loginErrorMessage.first()).toBeVisible({
      timeout: 10000,
    });
  }

  async verifyLoginPageIsDisplayed(): Promise<void> {
    await expect(this.userNameInput).toBeVisible();
    await expect(this.passwordInput).toBeVisible();
    await expect(this.loginBtn).toBeVisible();
  }
}